src/ray/common/scheduling/scheduling_alg.h

```diff
@@ -0,0 +1,147 @@
+#ifndef RAY_COMMON_SCHEDULING_SCHEDULING_H
+#define RAY_COMMON_SCHEDULING_SCHEDULING_H
+
+#include <iostream>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+
+/// Maximum number of predefined resources.
+#define NUM_PREDIFINED_RESOURCES 4
+/// List of predefined resources.
+/// The number of elements in enum should be equal to NUM_PREDIFINED_RESOURCES.
+enum PredefinedResources {CPU = 0, MEM = 1, GPU = 2, TPU = 3};
+
+/// Total and available capacity of a resource instance.
+struct ResourceCapacity {
+  int64_t total;
+  int64_t available;
+};
+
+/// Resource request.
+struct ResourceReq {
+  /// Amount of resource being requested.
+  int64_t demand;
+  /// Specify whether the request is soft or hard.
+  /// If hard, the entire request is denied if the demand exceeds the resource
+  /// availability. Otherwise, the request can be still be granted.
+  bool soft;
+};
+
+/// Resource request, including resource ID. This is used for custom resources.
+struct ResourceReqWithId {
```

Comment:
I think it would be better to just use a `std::pair` instead of defining a new struct for this.

Reply:
Could do this but I think it would make the code less readable. Using "first" and "second" instead of "id" and "req" seems less clear, no?